### PR TITLE
[WIP] liveblog element

### PR DESCRIPTION
**Pull request by collin**

### CODE REVIEW

File: elements/bulbs-liveblog/bulbs-liveblog.js [hunk 1]
```diff
@@ -0,0 +1,238 @@
...
+
+  getEntryIdsToFetch () {
+    let now = new Date();
+    let entryIds = [];
+    Object.keys(this.entriesData).forEach((entryId) => {
+      let entry = this.entriesData[entryId];
+      if (entry.published && entry.published < now && !this.entriesStore.all[entryId]) {
+        if (this.entriesStore.oldestEntryDate) {
+          if (entry.published >= this.entriesStore.oldestEntryDate) {
+            entryIds.push(entryId);
```
Comment (kand) [inline elements/bulbs-liveblog/bulbs-liveblog.js]: Can we tighten up the logic in these nested `if` statements? It's difficult to understand at a glance and confusing that two different branches end up in the same place.

File: elements/bulbs-liveblog/bulbs-liveblog.js [hunk 2]
```diff
@@ -0,0 +1,238 @@
...
+      this.resetSelectedEntry();
+    }
+  }
+
+  showNewEntries () {
+    this.removeNewEntriesButton();
+    while (this.entryStaging.firstElementChild) {
+      this.entriesContainer[0].prepend(this.entryStaging.firstElementChild);
+    }
+    let header = document.querySelector('header');
```
Comment (kand) [inline elements/bulbs-liveblog/bulbs-liveblog.js]: Could the header query selector be a configuration?

In it's current state, reliance on the `header` element is a hidden dependency, something that would be hard to find out if it wasn't working correctly for this reason.

If not a configuration, can this dependence be documented somewhere that's easily accessible if someone were to use this element?

File: elements/bulbs-liveblog/bulbs-liveblog.test.js [hunk 3]
```diff
@@ -0,0 +1,500 @@
...
+
+    });
+
+    xit('orders the firebase query by published', () => {
+
+    });
+
+    xit('limits the firebase query to 100 items', () => {
+
+    });
```
Comment (kand) [inline elements/bulbs-liveblog/bulbs-liveblog.test.js]: Are these tests still to be filled in? If not, can we remove them?

File: elements/bulbs-liveblog/bulbs-liveblog.test.js [hunk 4]
```diff
@@ -0,0 +1,500 @@
...
+  });
+
+  describe('setupInterval', () => {
+    it('connects an interval handler every LIVEBLOG_LATENCY seconds', () => {
+      sinon.stub(window, 'setInterval');
+      subject.setupInterval();
+      expect(window.setInterval).to.have.been.calledWith(subject.handleInterval, 5000);
+      window.setInterval.restore();
```
Comment (kand) [inline elements/bulbs-liveblog/bulbs-liveblog.test.js]: If you use a [sandbox](http://sinonjs.org/docs/#sandbox) you can restore in an `afterEach` for the whole suite:

```
var sandbox;

beforeEach(() => {
  sandbox = sinon.sandbox.create();
});

afterEach(() => {
  sandbox.restore();
});

it('should stub setInterval on the window', () => {
  sandbox.stub(window, 'setInterval');
  ...
});

```

Any place you would use `sinon` in the test suite, you now use `sandbox`.

If you get into the habit, it avoids potentially forgetting to restore a global and affecting the behavior of other tests. Also covers the case where you attempt to `restore` after a failed assertion, `restore` will never be called in the current case.

File: elements/bulbs-liveblog/bulbs-liveblog-entry.js [hunk 5]
```diff
@@ -0,0 +1,29 @@
+import { BulbsHTMLElement, registerElement } from 'bulbs-elements/register';
+import invariant from 'invariant';
+
+class BulbsLiveblogEntry extends BulbsHTMLElement {
+  attachedCallback () {
+    console.log('attachedCallback???');
```
Comment (collin) [inline elements/bulbs-liveblog/bulbs-liveblog-entry.js]: 👍

File: elements/bulbs-liveblog/bulbs-liveblog-entry.js [hunk 6]
```diff
@@ -0,0 +1,29 @@
...
+       '<bulbs-liveblog-entry> element MUST specify an `entry-id` attribute');
+
+    invariant(this.hasAttribute('entry-published'),
+       '<bulbs-liveblog-entry> element MUST specify an `entry-published` attribute');
+
+    this.dispatchEvent(new CustomEvent('liveblog-entry-attached'));
+  }
+
+  detachedCallback () {
+    console.log('detachedCallback!!!@');
```
Comment (collin) [inline elements/bulbs-liveblog/bulbs-liveblog-entry.js]: 👍

File: elements/bulbs-liveblog/bulbs-liveblog.js (hunk 1 above)
Comment (collin) [inline elements/bulbs-liveblog/bulbs-liveblog.js]:
```js
  getEntryIdsToFetch () {
    let now = new Date();
    let entryIds = [];
    Object.keys(this.entriesData).forEach((entryId) => {
      let entry = this.entriesData[entryId];
      let entryIsPublished = entry.published && entry.published < now;
      let entryIsCached = this.entriesStore.all[entryId];
      let entryIsOutOfBounds = entry.published
                                && this.entriesStore.oldestEntryDate
                                && this.entriesStore.oldestEntryDate > entry.published;

      if (entryIsPublished && !entryIsCached && !entryIsOutOfBounds) {
        entryIds.push(entryId);
      }
    });
    return entryIds;
  }
```

Does this read a little better?

File: elements/bulbs-liveblog/bulbs-liveblog.test.js (hunk 4 above)
Comment (collin) [inline elements/bulbs-liveblog/bulbs-liveblog.test.js]: 👍

File: elements/bulbs-liveblog/bulbs-liveblog.js [hunk 7]
```diff
@@ -0,0 +1,238 @@
...
+  handleEntryDetached () {
+    delete this.entriesStore.all[this.getAttribute('entry-id')];
+  }
+
+  getEntryIdsToFetch () {
+    let now = new Date();
+    let entryIds = [];
+    Object.keys(this.entriesData).forEach((entryId) => {
+      let entry = this.entriesData[entryId];
+      if (entry.published && entry.published < now && !this.entriesStore.all[entryId]) {
```
Comment (collin) [inline elements/bulbs-liveblog/bulbs-liveblog.js]: Can't think of a reason not to.

File: elements/bulbs-liveblog/bulbs-liveblog.js (hunk 2 above)
Comment (collin) [inline elements/bulbs-liveblog/bulbs-liveblog.js]: How would you want to configure this?

File: elements/bulbs-liveblog/bulbs-liveblog.js (hunk 1 above)
Comment (kand) [inline elements/bulbs-liveblog/bulbs-liveblog.js]: 👍

File: elements/bulbs-liveblog/bulbs-liveblog.js (hunk 2 above)
Comment (kand) [inline elements/bulbs-liveblog/bulbs-liveblog.js]: Not sure all the options available to us in this framework, but maybe an attribute with a default of `header`?

Comment (collin) [inline elements/bulbs-liveblog/bulbs-liveblog.js]: Like:
```html
<bulbs-liveblog scroll-offset-selector="#fancy-header">
```

File: elements/bulbs-liveblog/bulbs-liveblog.test.js (hunk 3 above)
Comment (collin) [inline elements/bulbs-liveblog/bulbs-liveblog.test.js]: These should be filled in, not sure I'll get to it today though.

**Pull request merged**